stan/math/rev/mat/functor/cvodes_integrator.hpp

```diff
@@ -0,0 +1,376 @@
+          cvode_state_sens_(NULL) {
+        stan::math::check_bounded("cvodes_integrator",
+                                  "solver", solver,
+                                  static_cast<size_t>(0),
+                                  static_cast<size_t>(2));
+
+        // Instantiate CVode memory
+        if (solver == 0) {
+          cvode_mem_ = CVodeCreate(CV_ADAMS, CV_FUNCTIONAL);
+
+          if (cvode_mem_ == NULL)
+            throw std::runtime_error("CVodeCreate failed to allocate memory");
+
+          set_cvode_options(rel_tol, abs_tol, max_num_steps);
+
+        } else if (solver == 1 || solver == 2) {
+          cvode_mem_ = CVodeCreate(CV_BDF, CV_NEWTON);
+
+          if (cvode_mem_ == NULL)
+            throw std::runtime_error("CVodeCreate failed to allocate memory");
+
+          set_cvode_options(rel_tol, abs_tol, max_num_steps);
+
+          // enable stability limit detection for solver==2
+          if (solver == 2)
+            check_flag(CVodeSetStabLimDet(cvode_mem_, 1), "CVodeSetStabLimDet");
+
+          // for the stiff solvers we need to reserve additional
+          // memory and provide a Jacobian function call
+          check_flag(CVDense(cvode_mem_, N_), "CVDense");
+          check_flag(CVDlsSetDenseJacFn(cvode_mem_, &ode::dense_jacobian),
+                      "CVDlsSetDenseJacFn");
+        }
+
+        // initialize forward sensitivity system of CVODES as needed
+        if (S_ > 0) {
+          cvode_state_sens_ = N_VCloneVectorArray_Serial(S_, cvode_state_);
+
+          for (size_t s = 0; s < S_ ; s++)
+            N_VConst(RCONST(0.0), cvode_state_sens_[s]);
+
+          // for the case with varying initials, the first N_
+          // sensitivity systems correspond to the initials which have
+          // as initial the identity matrix
+          if (initial_var_) {
+            for (size_t n = 0; n < N_; n++) {
+              NV_Ith_S(cvode_state_sens_[n], n) = 1.0;
+            }
+          }
+
+          check_flag(CVodeSensInit(cvode_mem_, static_cast<int>(S_),
+                                   CV_STAGGERED,
+                                   &ode::ode_rhs_sens, cvode_state_sens_),
+                     "CVodeSensInit");
+
+          check_flag(CVodeSensEEtolerances(cvode_mem_),
+                     "CVodeSensEEtolerances");
+        }
+      }
+
+      ~cvodes_integrator() {
+        // N_VDestroy_Serial should noop because
+        // N_VMake_Serial sets own_data to false
+        N_VDestroy_Serial(cvode_state_);
+        if (cvode_state_sens_ != NULL)
+          N_VDestroyVectorArray_Serial(cvode_state_sens_, S_);
+        CVodeFree(&cvode_mem_);
+      }
+
+      /**
+       * Forward the CVode array-based call from
+       * the ODE RHS to the Stan vector-based call
+       *
+       * @param[in] y state of the base ODE system
+       * @param[out] dy_dt ODE RHS at time t
+       * @param[in] t time
+       */
+      void rhs(const double y[], double dy_dt[], double t) const {
```

Comment:
Ok, done. Although this was inspired from how the call to the coupled_ode_system looks like...